uk_geo_utils/base_updater.py

```diff
@@ -0,0 +1,377 @@
+        return {
+            "constraint_name": constraint_name,
+            "temp_name": temp_name,
+            "constraintdef": constraintdef,
+            "temp_constraint_create_statement": f"ALTER TABLE {self.temp_table_name} ADD CONSTRAINT {temp_name} {constraintdef}",
+        }
+
+    def add_temp_primary_key(self):
+        self.stdout.write(f"Adding primary key to {self.temp_table_name}...")
+        self.stdout.write(
+            f"Executing: {self.primary_key_constraint['temp_constraint_create_statement']}"
+        )
+        self.cursor.execute(
+            self.primary_key_constraint["temp_constraint_create_statement"]
+        )
+
+    def get_foreign_key_constraints(self):
+        fkey_sql = f"""
+            SELECT conname AS constraint_name, confrelid::regclass::text AS refrenced_table, pg_get_constraintdef(oid), conrelid::regclass::text AS referencing_table
+            FROM pg_constraint 
+            WHERE contype = 'f'
+                AND (
+                    conrelid = '{self.table_name}'::regclass 
+                    OR
+                    confrelid = '{self.table_name}'::regclass
+                )      
+        """
+        self.cursor.execute(fkey_sql)
+        results = self.cursor.fetchall()
+
+        self.stdout.write(
+            f"Found {len(results)} foreign key constraints, where {self.table_name} is the referencing or referenced table"
+        )
+
+        fk_constraints = []
+
+        for row in results:
+            constraint_name = row[0]
+            # referenced_table_name = row[1]
+            constraintdef = row[2]
+            referencing_table = row[3]
+
+            fk_constraints.append(
+                {
+                    "constraint_name": constraint_name,
+                    "create_statement": f"ALTER TABLE {referencing_table} ADD CONSTRAINT {constraint_name} {constraintdef} NOT VALID",
+                    "delete_statement": f"ALTER TABLE {referencing_table} DROP CONSTRAINT IF EXISTS {constraint_name}",
+                    "validate_statement": f"ALTER TABLE {referencing_table} VALIDATE CONSTRAINT {constraint_name}",
+                }
+            )
+
+        return fk_constraints
+
+    def drop_foreign_keys(self):
+        self.stdout.write("Dropping foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            self.stdout.write(f"Executing: {constraint['delete_statement']}")
+            self.cursor.execute(constraint["delete_statement"])
+
+    def add_foreign_keys(self):
+        self.stdout.write("Creating foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            try:
+                self.stdout.write(
+                    f"Executing: {constraint['create_statement']}"
+                )
+                self.cursor.execute(constraint["create_statement"])
+            except (
+                ProgrammingError
+            ):  # better if psycopg2.errors.DuplicateObject:
```

Comment:
..and this one?

Reply:
Ah, this was a thing to come back to. I don't fully understand how the django cursor wraps psycopg2 in terms of error handling. The specific error psycopg2 error is `DuplicatedObject`, but what is actually raised by django is `ProgrammingError`. Ideally I'd only catch the `DuplicatedObject` ones, which I think can be done by inspecting the `Exception` object. I just didn't get to it.

Reply:
I _think_ if you catch a ProgrammingError `e`, have a look at

- `e.__cause__`
- `e.__cause__.pgerror`
- `e.__cause__.pgcode`

for more info about the wrapped exception

Reply:
:+1:  41d8893ba